tasks/indexing/task_pdp_indexing.go

```diff
@@ -0,0 +1,270 @@
+		}
+		return true, nil
+	}, harmonydb.OptionRetry())
+	if err != nil {
+		return xerrors.Errorf("committing transaction: %w", err)
+	}
+	if !comm {
+		return xerrors.Errorf("failed to commit transaction")
+	}
+
+	return nil
+}
+
+func (P *PDPIndexingTask) CanAccept(ids []harmonytask.TaskID, engine *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	// We just accept all tasks
+	// Note that this differs from markets v2 code which does a local storage check on the piece.
+	//
+	// If we are ever in an instance where the shared piece reader is expected to go to sector storage and will actually benefit from these local checks
+	// then we should be using markets v2 code because then we'll be in a shared datasource world.  In the rare case where an SP is handling PDP and PoRep on
+	// a cluster and just so happens to have a collision then the network overhead of moving a piece around will be small.
+	id := ids[0]
+	return &id, nil
+}
+
+func (P *PDPIndexingTask) TypeDetails() harmonytask.TaskTypeDetails {
+	return harmonytask.TaskTypeDetails{
+		Name: "PDPIndexing",
+		Cost: resources.Resources{
+			Cpu: 1,
+			Ram: uint64(P.insertBatchSize * P.insertConcurrency * 56 * 2),
+		},
+		Max:         P.max,
+		MaxFailures: 3,
+		IAmBored: passcall.Every(3*time.Second, func(taskFunc harmonytask.AddTaskFunc) error {
+			return P.schedule(context.Background(), taskFunc)
+		}),
+	}
+}
+
+func (P *PDPIndexingTask) schedule(_ context.Context, taskFunc harmonytask.AddTaskFunc) error {
+	// schedule submits
+	var stop bool
+	for !stop {
+		taskFunc(func(id harmonytask.TaskID, tx *harmonydb.Tx) (shouldCommit bool, seriousError error) {
+			stop = true // assume we're done until we find a task to schedule
+
+			var pendings []struct {
+				ID string `db:"id"`
+			}
+
+			err := tx.Select(&pendings, `SELECT id FROM pdp_piecerefs 
+            										WHERE indexing_task_id IS NULL
+            										AND needs_indexing = TRUE
+													ORDER BY created_at ASC LIMIT 1;`)
+			if err != nil {
+				return false, xerrors.Errorf("getting PDP pending indexing tasks: %w", err)
+			}
+
+			if len(pendings) == 0 {
+				return false, nil
+			}
+
+			pending := pendings[0]
+			_, err = tx.Exec(`UPDATE pdp_piecerefs SET indexing_task_id = $1 
+                             WHERE indexing_task_id IS NULL AND id = $2`, id, pending.ID)
+			if err != nil {
+				return false, xerrors.Errorf("updating PDP indexing task id: %w", err)
+			}
+
+			stop = false
+			return true, nil
+		})
+	}
+
+	return nil
+}
+
+func (P *PDPIndexingTask) Adder(taskFunc harmonytask.AddTaskFunc) {}
+
+var _ harmonytask.TaskInterface = &PDPIndexingTask{}
+var _ = harmonytask.Reg(&PDPIndexingTask{})
+
+func IndexCAR(r io.Reader, buffSize int, recs chan<- indexstore.Record, addFail <-chan struct{}) (int64, bool, error) {
+	blockReader, err := carv2.NewBlockReader(bufio.NewReaderSize(r, buffSize), carv2.ZeroLengthSectionAsEOF(true))
```

Comment:
```suggestion
	// ZeroLengthSectionAsEOF is not strictly needed here as it exists for the PoRep case where
	// padding pieces with zero bytes to get them to be a larger size is reasonable. This isn't
	// expected to be the case with PDP, but we'll stay consistent.
	blockReader, err := carv2.NewBlockReader(bufio.NewReaderSize(r, buffSize), carv2.ZeroLengthSectionAsEOF(true))
```

ZeroLengthSectionAsEOF was introduced specifically for this case, it's a weird feature that probably shouldn't exist. I wouldn't mind removing it from PDP but I can also imagine someone coming up with a whacky use for it... I wouldn't mind documenting it though for the future spelunker / maintainer.